api/src/main/java/org/pmiops/workbench/api/CloudTaskFreeCreditExpiry.java

```diff
@@ -0,0 +1,307 @@
+    final Set<DbUser> expiredUsers =
+        dbUsersWithChangedCosts.entrySet().stream()
+            .filter(
+                e ->
+                    CostComparisonUtils.costAboveLimit(
+                        e.getValue(),
+                        Math.max(
+                            dbCostByCreator.get(e.getKey()), liveCostByCreator.get(e.getKey())),
+                        workbenchConfig.get().billing.defaultFreeCreditsDollarLimit))
+            .map(Map.Entry::getValue)
+            .collect(Collectors.toSet());
+
+    final Set<DbUser> newlyExpiredFreeTierUsers = Sets.intersection(expiredUsers, freeTierUsers);
+
+    logger.info(
+        String.format(
+            "Found %d users exceeding their free tier limit, out of which, %d are new",
+            expiredUsers.size(), newlyExpiredFreeTierUsers.size()));
+
+    return newlyExpiredFreeTierUsers;
+  }
+
+  /**
+   * Find the users with db costs different from the live costs
+   *
+   * @param dbCostByCreator Map of userId->dbCost
+   * @param liveCostByCreator Map of userId->liveCost
+   * @return a {@link Map} of user Ids to their DbUsers
+   */
+  @NotNull
+  private Map<Long, DbUser> findDbUsersWithChangedCosts(
+      Map<Long, Double> dbCostByCreator, Map<Long, Double> liveCostByCreator) {
+    final Set<Long> usersWithChangedCosts =
+        liveCostByCreator.keySet().stream()
+            .filter(
+                u ->
+                    CostComparisonUtils.compareCosts(
+                            dbCostByCreator.get(u), liveCostByCreator.get(u))
+                        != 0)
+            .collect(Collectors.toSet());
+
+    logger.info(String.format("Found %d users with changed costs.", usersWithChangedCosts.size()));
+
+    Map<Long, DbUser> dbUsersWithChangedCosts =
+        StreamSupport.stream(userDao.findAllById(usersWithChangedCosts).spliterator(), false)
+            .collect(Collectors.toMap(DbUser::getUserId, Function.identity()));
+    return dbUsersWithChangedCosts;
+  }
+
+  private Set<DbUser> getFreeTierActiveWorkspaceCreatorsIn(Set<DbUser> users) {
+    return workspaceDao.findCreatorsByBillingStatusAndBillingAccountNameIn(
+        BillingStatus.ACTIVE,
+        new ArrayList<>(workbenchConfig.get().billing.freeTierBillingAccountNames()),
+        users);
+  }
+
+  private void deleteAppsAndRuntimesInFreeTierWorkspaces(DbUser user) {
+    logger.info("Deleting apps and runtimes for user " + user.getUsername());
+
+    impersonatedWorkspaceService.getOwnedWorkspaces(user.getUsername()).stream()
+        .map(WorkspaceResponse::getWorkspace)
+        .map(Workspace::getNamespace)
+        .map(workspaceDao::getByNamespace)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .filter(
+            dbWorkspace ->
+                workbenchConfig
+                    .get()
+                    .billing
+                    .freeTierBillingAccountNames()
+                    .contains(dbWorkspace.getBillingAccountName()))
```

Comment:
this part can be optimized. At least I see there are workspaceDao query twice. And I billingAccount might even already inside workspaceResponse? Can you check?